### ISSUE TITLE
String encoding has to be specified when reading dataset

### ISSUE THREAD

**Issue by matthew-d-jones** [ess-dmsc/h5cpp#410]
We are writing utf-8 encoded strings to file. To read such a dataset we seem to have to manually specify how it was encoded. This is not the case for ascii, which all the h5cpp unit tests we were using as documentation were using. Could it be handled automatically by h5cpp?

[Tested with 0.1.3 only]

This fails for utf-8:
```cpp
  auto StatusDataset = Group.get_dataset("alarm_status");
  std::vector<std::string> StatusData(StatusDataset.dataspace().size());
  StatusDataset.read(StatusData);
```
with a not particularly useful error:
```
HDF5-DIAG: Error detected in HDF5 (1.10.2) thread 0:
  #000: ../hdf5-1.10.2/src/H5Dio.c line 216 in H5Dread(): can't read data
    major: Dataset
    minor: Read failed
  #001: ../hdf5-1.10.2/src/H5Dio.c line 459 in H5D__read(): unable to set up type info
    major: Dataset
    minor: Unable to initialize object
  #002: ../hdf5-1.10.2/src/H5Dio.c line 1027 in H5D__typeinfo_init(): unable to convert between src and dest datatype
    major: Dataset
    minor: Feature is unsupported
  #003: ../hdf5-1.10.2/src/H5T.c line 4561 in H5T_path_find(): no appropriate function for conversion path
    major: Datatype
    minor: Unable to initialize object
```

This works:
```cpp
  auto StatusDataset = Group.get_dataset("alarm_status");
  std::vector<std::string> StatusData(StatusDataset.dataspace().size());
  auto Datatype = hdf5::datatype::String::variable();
  Datatype.encoding(hdf5::datatype::CharacterEncoding::UTF8);
  auto Dataspace = hdf5::dataspace::Simple({1});
  StatusDataset.read(StatusData, Datatype, Dataspace);
```

**Comment by eugenwintersberger** [comment]
I am currently on vacation but will have a look on this next week. If I
recall correctly, utf8 should be the default encoding. But I may be wrong.

Regards
Eugen

**Comment by zjttoefs** [comment]
Tests to be added in new test system

**Comment by jkotan** [comment]
Hi @matthew-d-jones,

In  the `issue_410` branch I've just added two tests i.e. 
```
H5pyStringCompatTest.test_read_scalar_string_utf8
H5pyStringCompatTest.test_read_scalar_string_simple_utf8
```
which tests your use-case. They work in v0.3.3.
I think the bug was fixed in v0.3.0 with other data types.
